### PR TITLE
[chore] [pkg/resourcetotelemetry] reduce memory usage

### PR DESCRIPTION
**Pull request by philchia**
Description:

1. Reduce memory usage by modify data in place, just let the pipeline builder to determine whether or not to clone data.

Link to tracking Issue:

N/A

Testing:

N/A

Documentation:

N/A

### CODE REVIEW

File: pkg/resourcetotelemetry/resource_to_telemetry.go [hunk 1]
```diff
@@ -33,8 +33,8 @@ func (wme *wrapperMetricsExporter) ConsumeMetrics(ctx context.Context, md pmetri
 }
 
 func (wme *wrapperMetricsExporter) Capabilities() consumer.Capabilities {
-	// Always return false since this wrapper clones the data.
-	return consumer.Capabilities{MutatesData: false}
+	// Always return true since this wrapper modify data inplace.
+	return consumer.Capabilities{MutatesData: true}
```
Comment (atoulme) [inline pkg/resourcetotelemetry/resource_to_telemetry.go:37]: this warrants a changelog imo

File: pkg/resourcetotelemetry/resource_to_telemetry.go [hunk 2]
```diff
@@ -107,6 +105,16 @@ func addAttributesToExponentialHistogramDataPoints(ps pmetric.ExponentialHistogr
...
+	var capacity = from.Len() + to.Len()
+	from.Range(func(k string, v pcommon.Value) bool {
+		_, ok := to.Get(k)
+		if ok {
+			capacity--
+		}
+		return true
+	})
+
+	to.EnsureCapacity(capacity)
```
Comment (atoulme) [inline pkg/resourcetotelemetry/resource_to_telemetry.go:117]: could you benchmark just this change separately to see if it helps at all with performance? This change could also be in its own PR

File: pkg/resourcetotelemetry/resource_to_telemetry.go [hunk 3]
```diff
@@ -33,8 +33,8 @@ func (wme *wrapperMetricsExporter) ConsumeMetrics(ctx context.Context, md pmetri
 }
 
 func (wme *wrapperMetricsExporter) Capabilities() consumer.Capabilities {
-	// Always return false since this wrapper clones the data.
-	return consumer.Capabilities{MutatesData: false}
+	// Always return true since this wrapper modify data inplace.
```
Comment (mx-psi) [inline pkg/resourcetotelemetry/resource_to_telemetry.go:36]:
```suggestion
	// Always return true since this wrapper modifies data inplace.
```

File: pkg/resourcetotelemetry/resource_to_telemetry.go (hunk 2 above)
Comment (philchia) [reply pkg/resourcetotelemetry/resource_to_telemetry.go]: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29430

File: .chloggen/optimize-resource-to-telemetry.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Not clone data in pkg/resourcetotelemetry by default
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [29327]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
```
Comment (mx-psi) [inline .chloggen/optimize-resource-to-telemetry.yaml:18]:
```suggestion
note: Do not clone data in pkg/resourcetotelemetry by default

# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
issues: [29327]

# (Optional) One or more lines of additional information to render under the primary note.
# These lines will be padded with 2 spaces and then inserted directly into the document.
# Use pipe (|) for multiline entries.
subtext: |
   - The resulting consumer will be marked as `MutatesData` instead
```

**Review by mx-psi — APPROVED** [review]

**Review by fatsheep9146 — APPROVED** [review]
